### PR TITLE
 Workaround for issue #2492 part 2 (improvement)

### PR DESCRIPTION
**Pull request by atamazov**
This is continuation of #2507.

- [debug] Disabled MIOPEN_DEBUG_WORKAROUND_ISSUE_2493 during driver warm-up.
  - Fixes small problem in [#2507](#2507) that affects the driver's warmup.
- [quality] Make the computation of max granularity loss more clear.
  - Resolves this [review comment from #2507](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2507#discussion_r1383815382).
  - The fix is in commit f186fafede484eff08277fa2392413720df3916c.
- [tests] test_db_sync:
  - Disabled `WORKAROUND_ISSUE_2493` via environment to get rid of false failures.
  - Added printing the number of failures per testing thread.
  - Some gfx1030-specific changes (all of them are DISABLED currently).
    - Reading legacy fdb (`WORKAROUND_ISSUE_1987`)
    - Skipping KDB and FDB testing (`SKIP_KDB_PDB_TESTING`).
    - W/A for ConvOclDirectFwdFused (`SKIP_CONVOCLDIRECTFWDFUSED`).

## :cyclone: Performance testing results

Preconditions:
- Navi21 (gfx1030/36)
- ROCm 5.7 docker, Ubuntu 20.04.5 LTS
- BUILD_DEV=On, Release build
- branch 6530353d8bed4f6dd96f2bda8231feb9c9e122d2 vs. baseline `develop` 19c6046d1a1d2ac7123fbeaf7775dd7a5749bfd3
- 93 known FP32 convolution configs

Tested modes:
- Immediate mode with default (system) find-db

Bottom line: No performance or correctness regressions on average. Detailed logs & csv files available upon request.

----
[Attribution] @junliume @JehandadKhan
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/quality
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/debugging
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/urgency_high
- https://github.com/ROCmSoftwarePlatform/MIOpen/labels/value_high (fixes warmup in the driver that affects performance testing)
- Proposed reviewers:
  - @JehandadKhan
  - @CAHEK7

### CODE REVIEW

File: src/solver/conv_winoRxS.cpp [hunk 1]
```diff
@@ -682,10 +682,14 @@ static bool IsApplicableBase(const ExecutionContext& ctx, const ProblemDescripti
         // clang-format on
 
 #if WORKAROUND_ISSUE_2493
-    if(!miopen::IsDisabled(MIOPEN_DEBUG_WORKAROUND_ISSUE_2493{}))
+    if(!miopen::IsDisabled(MIOPEN_DEBUG_WORKAROUND_ISSUE_2493{}) && !miopen::debug::IsWarmupOngoing)
```
Comment (JehandadKhan) [inline src/solver/conv_winoRxS.cpp:685]: Why is the rationale behind this?

**Review by JehandadKhan — DISMISSED** [review]

**Comment by junliume** [comment]
@cderb @JehandadKhan `dbsync gfx90a` has failed but I think it is expected? KDB update might be required for this PR.

**Comment by atamazov** [comment]
@junliume I'll look into into this; should be easy to fix.

File: src/solver/conv_winoRxS.cpp (hunk 1 above)
Comment (atamazov) [reply src/solver/conv_winoRxS.cpp:685]: @JehandadKhan Without this fix, Winograd assembling is skipped during warm-up and that leads to inaccurate host-side performance testing results. Specifically, I see false ~36% drop for Average Aux Wall time (Forward) after merging #2507.

**Comment by atamazov** [comment]
@junliume @JehandadKhan @shurale-nkn The root reason of the issue is that before merging, PR #2492 wasn't updated from `develop` and then tested on CI. The immediate reason is that `dbsync` tests require `MIOPEN_DEBUG_WORKAROUND_ISSUE_2493=0` in the environment. I'll prepare the fix and let you know. Most likely, I'll make it right in this PR.

**Comment by atamazov** [comment]
It wasn't that easy as test was aborting on gfx1030.

File: test/gtest/db_sync.cpp [hunk 2]
```diff
@@ -254,8 +304,19 @@ void ParseFDBbVal(const std::string& val, std::vector<FDBVal>& fdb_vals)
...
+        auto id     = id_val.substr(0, id_size);
+        auto values = id_val.substr(id_size + 1);
+#if WORKAROUND_ISSUE_1987
+        /// \todo Copied from src/db_record.cpp
+        /// Detect legacy find-db item (v.1.0 ID:VALUES) and transform it to the current format.
+        /// For now, *only* legacy find-db record use convolution algorithm as ID, so if ID is
+        /// a valid algorithm, then we can safely assume that the item is in legacy format.
+        if(IsValidConvolutionDirAlgo(id))
+        {
+            ASSERT_TRUE(TransformFindDbItem10to20(id, values))
```
Comment (atamazov) [inline test/gtest/db_sync.cpp:318]: Why this is needed: gfx1030 fdb is in legacy format.

File: test/gtest/db_sync.cpp [hunk 3]
```diff
@@ -562,6 +628,21 @@ void CheckFDBEntry(size_t thread_index,
...
         {
             miopen::solver::Id id{val.solver_id};
+            EXPECT_TRUE(id.IsValid())
+                << '[' << (++failures) << "] " //
+                << "Solver " << id.Value() << "/" << id.ToString() << ", val.solver_id "
+                << val.solver_id << ", val.vals " << val.vals;
+
+            /// \todo Workaround: solv.IsApplicable() asserts with ConvOclDirectFwdFused
+            /// on gfx1030. AnySolver instance is empty (nullptr) due to some unknown reason.
+            if(val.solver_id == "ConvOclDirectFwdFused")
```
Comment (atamazov) [inline test/gtest/db_sync.cpp:641]: Strange problem that we can easily postpone for a while.

File: test/gtest/db_sync.cpp [hunk 4]
```diff
@@ -550,6 +615,7 @@ void CheckFDBEntry(size_t thread_index,
         problem.Serialize(ss);
         // moment of truth
         EXPECT_TRUE(ss.str() == kinder.first)
+            << '[' << (++failures) << "] " //
```
Comment (atamazov) [inline test/gtest/db_sync.cpp:620]: This prints number of failures in the current thread.

**Review by atamazov — COMMENTED** [review]
Tips for reviewers

**Comment by atamazov** [comment]
@junliume @JehandadKhan @CAHEK7 Made some fixes that should unblock CI. Tested on gfx1030. Some [hints for reviewers](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2510#pullrequestreview-1721406568) provided. Ready for the next review & CI testing round.

**Review by junliume — APPROVED** [review]
dbsync tests are passing now
